src/Service/GatewayResourceService.php

```diff
@@ -142,33 +142,48 @@ public function getSource(string $reference, string $pluginName): ?Source
     }//end getSource()
 
     /**
-     * Find all sources that have a location that match the specified url.
+     * Find a source that has a location that match the specified url.
      * Todo: we should use a mongoDB filter instead of this, sources should exist in MongoDB.
-     * Todo: for future reference, there is a function with very similar BL in the CustomerInteractionBundle->CustomerInteractionService->getSource(), we could/should merge that code with this function.
+     * This function gets used in the CustomerInteractionBundle, Gateway->synchronizationService.
      *
-     * @param string $url        The url we are trying to find a matching source for.
-     * @param string $pluginName The name of the plugin that requests these resources.
+     * @param string      $url        The url we are trying to find a matching source for.
+     * @param string      $pluginName The name of the plugin that requests these resources.
+     * @param string|null $endpoint   The resulting endpoint (the remainder of the path).
      *
-     * @return array|null
+     * @return Source|null The source found or null.
      */
-    public function findSourcesForUrl(string $url, string $pluginName): ?array
+    public function findSourceForUrl(string $url, string $pluginName, ?string &$endpoint = null): ?Source
     {
-        $sources    = [];
-        $allSources = $this->entityManager->getRepository('App:Gateway')->findAll();
-
-        foreach ($allSources as $source) {
-            if (empty($source->getLocation()) === false && str_contains($url, $source->getLocation()) === true) {
-                $sources[] = $source;
+        // 1. Get the domain from the url
+        $parse    = \Safe\parse_url($url);
+        $location = $parse['scheme'].'://'.$parse['host'];
+
+        // 2.a Try to establish a source for the domain
+        $source = $this->entityManager->getRepository('App:Gateway')->findOneBy(['location' => $location]);
+
+        // 2.b The source might be on a path e.g. /v1 so if whe cant find a source let try to cycle
+        if ($source instanceof Source === false && isset($parse['path']) === true) {
```

Comment:
Shouldn't we use === null instead of instanceOf

Reply:
Good point, i copy pasted this from somewhere else without changing style stuff😝